ortools/linear_solver/xpress_interface.cc

```diff
@@ -1925,11 +1895,49 @@ MPSolver::ResultStatus XpressInterface::Solve(MPSolverParameters const& param) {
   return result_status_;
 }
 
+namespace {
+  template<class T>
+  // T = MPVariable | MPConstraint
+  // or any class that has a public method name() const
+  void ExtractNames(XPRSprob mLp, const std::vector<T*>& objects, int type) {
```

Comment:
i think we can deduce type from T t...t same

Reply:
Ideally, we could add a static public member `type` to classes `MPConstraint` & `MPVariable`, then use `T::type`. But this is only used for XPRESS, so it would be a bit awkward.

Reply:
template <typename T, class std::enable_if_t<std::is_same<T, MPConstraint>, std::is_same<T, MPVariable> >::type>

Reply:
And you may use std::is_same inside the function to recover the type.

Reply:
`#include <iostream>

using namespace std;
struct A{};
struct B{};
struct C{};
template <typename T, class = typename std::enable_if_t<
                          std::is_same<A, T>::value ||
                          std::is_same<B, T>::value>>
          void f(T a){}
int main()
{
    A a;
    C c;
    f<A>(a);
    // will not compile
    //f<C>(c);

    return 0;
}`